Applications/Spire/Source/Ui/EditableTableView.cpp

```diff
@@ -0,0 +1,431 @@
+  TableModelTransactionLog m_transaction;
+  scoped_connection m_source_connection;
+
+  explicit RevertColumnTableModel(std::shared_ptr<TableModel> source)
+    : m_source(std::move(source)),
+      m_source_connection(m_source->connect_operation_signal(
+        std::bind_front(&RevertColumnTableModel::on_operation, this))) {}
+
+  int get_row_size() const {
+    return m_source->get_row_size();
+  }
+
+  int get_column_size() const {
+    return m_source->get_column_size() - 2;
+  }
+
+  AnyRef at(int row, int column) const {
+    return m_source->at(row, column + 1);
+  }
+
+  QValidator::State set(int row, int column, const std::any& value) {
+    return m_source->set(row, column + 1, value);
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const {
+    return m_transaction.connect_operation_signal(slot);
+  }
+
+  void on_operation(const TableModel::Operation& operation) {
+    auto adjust_row = [] (int index, const AnyListModel& source) {
+      auto row = std::make_shared<ArrayListModel<std::any>>();
+      for(auto i = 1; i < source.get_size() - 1; ++i) {
+        row->push(source.get(i));
+      }
+      return row;
+    };
+    visit(operation,
+      [&] (const TableModel::AddOperation& operation) {
+        m_transaction.push(TableModel::AddOperation(operation.m_index,
+          adjust_row(operation.m_index, *operation.m_row)));
+      },
+      [&] (const TableModel::MoveOperation& operation) {
+        m_transaction.push(operation);
+      },
+      [&] (const TableModel::RemoveOperation& operation) {
+        m_transaction.push(TableModel::RemoveOperation(operation.m_index,
+          adjust_row(operation.m_index, *operation.m_row)));
+      },
+      [&] (const TableModel::UpdateOperation& operation) {
+        m_transaction.push(TableModel::UpdateOperation(operation.m_row,
+          operation.m_column - 1, operation.m_previous, operation.m_value));
+      });
+  }
+};
+
+struct EditableTableModel : TableModel {
+  std::shared_ptr<TableModel> m_source;
+  std::shared_ptr<EditableTableView::HeaderModel> m_header;
+  TableModelTransactionLog m_transaction;
+  scoped_connection m_source_connection;
+
+  explicit EditableTableModel(std::shared_ptr<TableModel> source,
+    std::shared_ptr<EditableTableView::HeaderModel> header)
+    : m_source(std::move(source)),
+      m_header(std::move(header)),
+      m_source_connection(m_source->connect_operation_signal(
+        std::bind_front(&EditableTableModel::on_operation, this))) {}
+
+  int get_row_size() const {
+    return m_source->get_row_size() + 1;
+  }
+
+  int get_column_size() const {
+    return m_header->get_size() + 2;
+  }
+
+  AnyRef at(int row, int column) const {
+    if(row < 0 || row >= get_row_size() || column < 0 ||
+        column >= get_column_size()) {
+      throw std::out_of_range("The row or column is out of range.");
+    }
+    if(column == 0) {
+      static auto row_index = 0;
```

Comment:
`AnyRef(row, AnyRef::by_value)` doesn’t work for `TableModel::get`. `AnyRef(row, AnyRef::by_value)` is returned from `EditableTableModel::at`, but it is destroyed before returning from `TableModel::get`, as a result, `TableModel::get` returns an invalid value.

```
template<typename T>
const T& TableModel::get(int row, int column) const {
  return any_cast<const T>(at(row, column));
}
```

To avoid it, I used `TableModel::at ` directly instead of `TableModel::get`:
```
m_rows.set(any_cast<int>(m_table_body->get_table()->at(i, 0)), row);
```